include/etcpal/cpp/error.h

```diff
@@ -70,125 +75,852 @@ namespace etcpal
 ///   // Prints "Error: 'Out of memory'"
 /// }
 /// \endcode
+///
+/// Result is often used as the return type for functions; it has an implicit constructor from
+/// #etcpal_error_t to simplify return statements:
+/// \code
+/// etcpal::Result DoSomething()
+/// {
+///   // ... After we did something successfully...
+///   return kEtcPalErrOk;
+/// }
+/// \endcode
 class Result
 {
 public:
   Result() = delete;
   // Note: this constructor is not explicit by design, to allow implicit conversions e.g.
-  //   etcpal::Result res = etcpal_c_function_that_returns_error_t();
-  Result(etcpal_error_t code);
-  Result& operator=(etcpal_error_t code);
+  //   etcpal::Result FunctionReturningResult()
+  //   {
+  //     return kEtcPalErrOk;
+  //   }
+  constexpr Result(etcpal_error_t code) noexcept;
+  Result& operator=(etcpal_error_t code) noexcept;
 
-  bool IsOk() const;
-  etcpal_error_t code() const;
+  constexpr bool IsOk() const noexcept;
```

Comment:
I clearly have to wrap my head around constexpr more.  The awesome fact that simple state can be computed at compile time..